Special-case `NumberType` for backwards compatibilty

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61533**

Differential Revision: [D29661173](https://our.internmc.facebook.com/intern/diff/D29661173)

Commit:
Special-case `NumberType` for backwards compatibilty

[ghstack-poisoned]

Commit:
Update on "Special-case `NumberType` for backwards compatibilty"




[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29661173).

Commit:
Update on "Special-case `NumberType` for backwards compatibilty"


Differential Revision: [D29661173](https://our.internmc.facebook.com/intern/diff/D29661173)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29661173).

Commit:
Update on "Special-case `NumberType` for backwards compatibilty"


Differential Revision: [D29661173](https://our.internmc.facebook.com/intern/diff/D29661173)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29661173).

Commit:
Update on "Special-case `NumberType` for backwards compatibilty"


Differential Revision: [D29661173](https://our.internmc.facebook.com/intern/diff/D29661173)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29661173).

tugsbayasgalan commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1410,11 +1414,13 @@ struct to_ir {
-            k->type()->isSubtypeOfExt(dict_type_hint->getKeyType(), &ss);
+            k->type()->isSubtypeOfExt(dict_type_hint->getKeyType(), &ss) ||
+            (k->type() == NumberType::get() &&
+             dict_type_hint->getKeyType()->isSubtypeOfExt(k->type(), &ss));
```
Just making sure. Are we saying int is a subtype of Number?

tugsbayasgalan replied:
If the Number type is actually not an integer, wouldn't it fail?

Commit:
Update on "Special-case `NumberType` for backwards compatibilty"


Differential Revision: [D29661173](https://our.internmc.facebook.com/intern/diff/D29661173)

[ghstack-poisoned]

Commit:
Update on "Special-case `NumberType` for backwards compatibilty"


Differential Revision: [D29661173](https://our.internmc.facebook.com/intern/diff/D29661173)

[ghstack-poisoned]

Commit:
Update on "Special-case `NumberType` for backwards compatibilty"


Differential Revision: [D29661173](https://our.internmc.facebook.com/intern/diff/D29661173)

[ghstack-poisoned]

tugsbayasgalan commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3899,11 +3912,12 @@ struct to_ir {
+                      types[i] == NumberType::get(),
                   "Type "
-                  "hint for dict was",
+                  "hint for dict was ",
```
#grammar :)

gmagogsfm commented:
I think based on our offline discussion, this is no longer needed because the broken test has been fixed. Closing this PR.
